gapic/src/main/com/google/gapid/perfetto/models/BatterySummaryTrack.java

```diff
@@ -139,26 +159,187 @@ private String counterSQL() {
     public final long[] ts;
     public final long[] capacity;
     public final long[] charge;
     public final long[] current;
 
-    public Data(DataRequest request, long[] ts, long[] capacity, long[] charge, long[] current) {
+    public Data(DataRequest request, long[] id, long[] ts, long[] capacity, long[] charge,
+        long[] current) {
       super(request);
+      this.id = id;
       this.ts = ts;
       this.capacity = capacity;
       this.charge = charge;
       this.current = current;
     }
 
     public static Data empty(DataRequest req) {
-      return new Data(req, new long[0], new long[0], new long[0], new long[0]);
+      return new Data(req, new long[0], new long[0], new long[0], new long[0], new long[0]);
+    }
+  }
+
+  public static class Values implements Selection<Long>, Selection.Builder<Values> {
+    public final long[] ts;
+    public final long[] dur;
+    public final long[] capacity;
+    public final long[] charge;
+    public final long[] current;
+    private final Set<Long> valueKeys;
+
+    public Values(long[] ts, long[] dur, long[] capacity, long[] charge, long[] current,
+        Set<Long> valueKeys) {
+      this.ts = ts;
+      this.dur = dur;
+      this.capacity = capacity;
+      this.charge = charge;
+      this.current = current;
+      this.valueKeys = valueKeys;
+    }
+
+    @Override
+    public String getTitle() {
+      return "Batteries";
+    }
+
+    @Override
+    public boolean contains(Long key) {
+      return valueKeys.contains(key);
+    }
+
+    @Override
+    public Composite buildUi(Composite parent, State state) {
+      return new BatterySelectionView(parent, state, this);
+    }
+
+    @Override
+    public Selection.Builder<Values> getBuilder() {
+      return this;
+    }
+
+    @Override
+    public void getRange(Consumer<TimeSpan> span) {
+      long min = Arrays.stream(ts).min().getAsLong();
+      long max = IntStream.range(0, ts.length).mapToLong(i -> ts[i] + dur[i]).max().getAsLong();
+      span.accept(new TimeSpan(min, max));
+    }
+
+    @Override
+    public Values combine(Values that) {
+      if (ts.length == 0) {
+        return that;
+      } else if (that.ts.length == 0) {
+        return this;
+      }
+
+      long[] newTs = new long[this.ts.length + that.ts.length];
+      long[] newDur = new long[this.ts.length + that.ts.length];
+      long[] newCapacity = new long[this.ts.length + that.ts.length];
+      long[] newCharge = new long[this.ts.length + that.ts.length];
+      long[] newCurrent = new long[this.ts.length + that.ts.length];
+
+      int ai = 0, bi = 0, ri = 0;
+      for (; ai < this.ts.length && bi < that.ts.length; ri++) {
+        long at = this.ts[ai], bt = that.ts[bi];
+        boolean selectThis = true;
+        if (at == bt) {
+          ai++;
+          bi++;
+        } else if (at < bt) {
+          ai++;
+        } else {
+          bi++;
+          selectThis = false;
+        }
+        if (selectThis) {
```

Comment:
Is there a good reason behind choosing arrays for the members of Values? I've seen this in all the tracks. The logic here is the same as merge sort merging phase, which I feel can be much simplified if we use containers instead of arrays. I'm not being very picky here, but just curious.

Reply:
I think what you says makes great sense! I wrote it basically following the existing patterns in AGI, especially the counter track. 
After a closer look I believe Pascal's original choice for array was to lend convenience to CounterSelectionView. With array, each array can directly act as a real column in the selection info window, with an elegant implementation using `createTableColumn` and `setInput` . A tradeoff for simplicity here and the simplicity there have to be made...

Reply:
+1, we really should't be re-inventing the wheels like this.